TST: Adding testing for lapack gttrf and gttrs

swallan opened the pull request:
This commit begins the tests for lapack gttrf and gttrs
@mdhaber
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message: TST: Adding testing for lapack gttrf and gttrs
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Adds tests for #11123: Add wrapper for ?gttrf/?gttrs 

#### What does this implement/fix?
<!--Please explain your changes.-->
Adds tests for lapack gttrf and gttrs 
#### Additional information
<!--Any additional information you think is important.-->

One thing that I had some trouble with was verifying that b was not being overridden in gttrs. It appeared to me that only in dtypes np.float32 and np.complex64 was it being maintained, not np.float64 or np.complex128. Maybe this hasn't been implemented yet as I didn't see the same ```(in, out, copy)``` for b in gttrs. I’m still working on covering all testing bases.

Kai-Striega committed:
ENH: linalg: Add bindings for ?gttrf

Kai-Striega committed:
ENH: linalg: Add bindings for ?gttrs

Kai-Striega committed:
ENH: linalg: Allow copies for ?gttrf

This commit allows ?gttrf to copy the input arrays dl, d and du. This
ensures the array's original content is preserved, rather than the
behaviour seen in LAPACK, which is to overwrite the array. The choice to
copy by default was made to keep the behaviour consistent with ?gtsv.

swallan committed:
TST: Adding testing for lapack gttrf and gttrs

This commit begins the tests for lapack gttrf and gttrs

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+    np.random.seed(42)
+    n = 10
+    for index, dtype in enumerate(DTYPES):
+        atol = 10**-(np.finfo(dtype).precision-1)
```
I'd suggest changing `rtol`, instead. For instance,
```
rtol=25*np.finfo(dtype).eps
```
appears 12 times in https://github.com/scipy/scipy/blob/master/scipy/linalg/tests/test_lapack.py.

I think it makes more sense because ultimately the reason we have to change a tolerance is that the _relative_ tolerance of single precision arithmetic is worse than double precision arithmetic.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+            d = np.random.rand(n)
+            dl = np.random.rand(n-1)
+        else:
+            du = np.random.rand(n-1) * 1j
```
Let's make these complex, not just imaginary, e.g.
```suggestion
            du = np.random.rand(n-1) + np.random.rand(n-1) * 1j
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+
+        _dl, _d, _du, du2, ipiv, info = dgttrf(dl, d, du)
+
+        np.testing.assert_allclose(dl, diag_cpy[0], atol=atol)
```
I would `assert_array_equal` here, right? We want to verify that they haven't changed _at all_.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+        b_cpy = b.copy()
+        x_dgttrs, info = dgttrs(_dl, _d, _du, du2, ipiv, b)
+        np.testing.assert_allclose(x, x_dgttrs, atol=atol)
+        np.testing.assert_allclose(b, b_cpy, atol=atol)
```
Again, `b` and `b_copy` should be identical.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+        np.testing.assert_allclose(b, b_cpy, atol=atol)
+        
+
+        dl_malformatted = np.append(dl, [1])
```
I wouldn't call it malformatted. It's the wrong shape.
Also, I wouldn't use `np.append(dl, [1])`. Probably better to just re-use `d` here, or maybe `np.random.rand(n)`.

swallan replied:
I'll change it to `dl_misshaped,` `d_misshaped,` and `du_misshaped` with a new random matrices.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+
+        dl_malformatted = np.append(dl, [1])
+        d_malformatted = np.append(d, [1])
+        du_malformatted = np.append(du, [1])
```
Doesn't need to be different from the one used for `dl`.

mdhaber commented:
@swallan nice start.
@Kai-Striega, I think @swallan is right that `?gttrs` needs to copy `b` and have an `overwrite_b` argument.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+        with assert_raises(ValueError):
+            dgttrf(dl, d_malformatted, du)
+        with assert_raises(ValueError):
+            dgttrf(dl, d, du_malformatted)
```
Let's also see what happens when n = 2. @Kai-Striega has explicitly included a check in the wrapper:
`integer intent(hide),   depend(d), check(len(d) > 2) :: n = len(d)`
We should make sure it's working.

swallan replied:
It does does reject it and throws a nasty error that I can't seem to catch even with the most general: `assert_raises(Exception)`. <a href="https://ibb.co/NZ85GM1"><img src="https://i.ibb.co/Qr2Lx48/Screen-Shot-2019-12-15-at-3-33-49-PM.png" alt="Screen-Shot-2019-12-15-at-3-33-49-PM" border="0"></a>

Kai-Striega replied:
We cannot have n = 2 to ensure `du2`, an array of dimension ``n-2``, keeps a positive dimension.  You should still be able to catch the exception. Could you look at the test for `?tzrzf` - this tests a similar case for a different function, does that test pass for you? 

@mdhaber As an aside I do think we should expose this as some kind of more specialised Exception Class. From a user's perspective having to catch ``Exception``s is usually not the best way to go.

swallan replied:
You're right, it is catchable. Must have been a type with what I had before.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+        dgttrs = get_lapack_funcs('gttrs', dtype=dtype)
+
+        _dl, _d, _du, du2, ipiv, info = dgttrf(dl, d, du)
+
```
Let's add the test that the factors multiply to reproduce the (permuted) original matrix.

swallan replied:
Haven't finished this yet. It should be as simple as what we did to permuting A to get the new factors, yes? (Like we did in an earlier meeting)

swallan replied:
@mdhaber What factor is this missing from this? (At lines 1681-1686 if they don't display above). I think that the permutation order is being done correctly on the original matrix, but permuted A is not matching the combined dl, d, du outputs of gttrf. Is there additional swapping / eliminations that are being done outside of IPIV with du2?

mdhaber replied:
The input `dl`, `d`, `du` describe the original matrix `A` as you have written, `np.diag(dl, -1) + np.diag(d, 0) + np.diag(du, 1)`.
The output variables `_dl`, `_d`, `_du`, and `_du2` also describe `A` but in a different way. 
`_dl` is the -1 diagonal of a unit-diagonal, lower triangular matrix `L`: `L = np.eye(n) + np.diag(dl, -1)`.
`_d`, `_du`, and `_du2` are the first three diagonals of an upper triangular matrix `U`: `U = np.diag(_d) + np.diag(_du, 1) + np.diag(_du2, 2)`.
Permuted `A` is the product of `L` and `U`, so `assert_allclose(L@U, A[perm_fnl], rtol=rtol)`.
I think.

swallan replied:
Hmm, I am seeing negative numbers inside the result of L@U despite A originally having none. I let L be `np.eye(n) + np.diag(_dl, -1)` and U be `np.diag(_d, 0) + np.diag(_du, 1) + np.diag(du2, 2)`, but `L@U` gives a couple values that are negative, so even before permuting it seems that something still isn't right here.

I saw some talk of multipliers on the  [NAG Manual example result](https://www.nag.com/numeric/fl/nagdoc_latest/examples/baseresults/f07cdfe.r.html) page, are those relevant here?

Not much new code other than what's above, so I didn't add any commits for this.

mdhaber replied:
I see. Well, I'm pretty certain that `U` is right because the documentation is very explicit, e.g. 

> On exit, D is overwritten by the n diagonal elements of the upper triangular matrix U from the LU factorization of A.

and that's what `U = np.diag(_d)...` does.

But the documentation about `dl` does not say "subdiagonal" of `L`:

> On exit, DL is overwritten by the (n-1) multipliers that define the matrix L from the LU factorization of A.

I assumed it meant subdiagonal, but it says "multipliers".
This explains it: http://www.hpcavf.uclan.ac.uk/softwaredoc/sgi_scsl_html/sgi_html/ch03.html
but it's not 100% explicit about exactly how to form `L` from the multipliers; the factored form of `dgttrf` is different than that for the example `dsytrf`.

I tried out some things and here is what I found that works:
```
n = 10
dl = np.random.rand(n-1)
d = np.random.rand(n)
du = np.random.rand(n-1)
x = np.random.rand(n)

A = np.diag(d) + np.diag(dl, -1) + np.diag(du, 1)
b = A@x
_dl, _d, _du, _du2, ipiv, info = dgttrf(dl, d, du)

perm_fnl = np.arange(n)
for i, piv in enumerate(ipiv):
    perm_fnl[i], perm_fnl[piv-1] = perm_fnl[piv-1], perm_fnl[i]

U = np.diag(_d) + np.diag(_du, 1) + np.diag(_du2, 2)
L = np.eye(n)

for i, m in enumerate(_dl):
    piv = ipiv[i] - 1

    # create permutation matrix
    P = np.eye(n)
    temp = P[i,:].copy()
    P[i,:] = P[piv, :]
    P[piv,:] = temp

    # create factor Li
    Li = np.eye(n)
    Li[i+1,i] = m

    L = L@P@Li

# one last permutation
piv = ipiv[-1] - 1
i = -1
P = np.eye(n)
temp = P[i,:].copy()
P[i,:] = P[piv, :]
P[piv,:] = temp

print(A - L@U)
```
It's horrendously inefficient multiplying all those zeros. As a matter of form, I'd suggest eliminating those explicit permutation matrix multiplications and replacing them with explicit column swaps. (Think about what the permutation matrix does, and do just that.) Similarly, you can eliminate the explicit `Li` matrices, replacing them with elementary column operations.

Alternatively, instead of doing all this `P1@L1@...@Pn-1@Ln-1@Pn` to generate `L` and then performing `L@U`, you can just do it all in reverse on `U`: a sequence of row permutations and elementary row operations. 

But I think the first approach - generating `L` explicitly - is better for the test because it's a little more explicit that this indeed just an `LU` decomposition.

swallan replied:
I see. I read that article, and even though I'm not 100% on it, I think I see the general idea of what is going on with what you provided. It looks like A - L@U brings all of the entries in the matrix to near zero. 

However, I've been running into an issue with how assert_equals compares A and L@U. For example, if the two elements from the matrix being compared were` 0.` and `6.059189e-09, ` with the way troll works, I think that it won't end up passing as true. I saw this: `absolute(a - b) <= (atol + rtol * absolute(b))` in the documentation, so I thought that possibly the problem would be `b` was ending up as zero and therefore the tolerance was zero. But, I think since either `a` or `b` has to be zero, modifying rtol here might not work. Is it possible that I can change atol instead since it it value independent?

I'm working on adapting it to work within the for loop for complex numbers as well.

mdhaber replied:
Yes, it's ok to change `atol` too, if you need to. But how big is `atol`? In your situation, if `atol > 6.059189e-09`, it will pass, right? And I think `atol` is generally pretty loose.

mdhaber replied:
It seems that `atol` was 0, which is why this wasn't passing. Go ahead and set a reasonable `atol`. `1e-7` worked for me.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
+            dgttrf(dl, d_malformatted, du)
+        with assert_raises(ValueError):
+            dgttrf(dl, d, du_malformatted)
+
```
Please also add a test for what happens when the input matrix is singular (e.g. one row is all zeros). The routine should fail gracefully, perhaps returning nonzero `info`.

swallan replied:
It currently does not, but I added a test for that when d is all zeroes.

Kai-Striega replied:
@swallan if it doesn't fail gracefully then how does it fail?

swallan commented:
Additionally, since we have the `overwrite_*=0` argument, if I set it to 1 it should overwrite, correct?

mdhaber commented:
> Additionally, since we have the overwrite_*=0 argument, if I set it to 1 it should overwrite, correct?

Yes, but I don't think you have to test for it.

swallan commented:
Ok. I ask since I did check that and it seems that they're not always overwritten when `overwrite_dl=1`.

mdhaber commented:
I was wondering about that actually. In several places in SciPy documentation, arguments like `overwrite_b` describe it like `b` _may_ be overwritten; it sounds like it's not guaranteed. Usually the reason to use the options is the possibility of slightly better speed (no copying), not because you _want_ the data to be overwritten. Anyway, this must just be how f2py does things, so I wouldn't worry about it.

swallan committed:
TST: Add additional tests for ?gttrf and ?gttrs

Adds tests for singular and 2x2 matrices. Uses rtol instead of atol. Imaginary->complex numbers. Clarifies use of assert_allclose to assert_array_equal for arrays that should be completely unchanged.

swallan commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1672,23 +1675,39 @@ def test_gttrf_gttrs():
+        #               "?gttrf: singular matrix should return non-zero info.")
+        
+        # put in a 2x2 matrix:    
+        # dgttrf(dl[0], d[:1], du[0])
```
I decided to leave these commented out as they will fail if run right now.

mdhaber replied:
How do they fail?

swallan replied:
Oops, should have elaborated more on that. When inputting a singular matrix, it doesn't return non-zero info, so I wrote an assert_() to expect that in the future.  

The 2x2 matrix input does raise an error, but I initially had some trouble getting it to be caught when using `with assert_raises(Exception):`. That's fixed now, though. Not sure what the issue was but I assume some small typo.

mdhaber replied:
> When inputting a singular matrix, it doesn't return non-zero info

What _does_ it do? Does it just output an incorrect solution?

swallan replied:
My mistake–it does return a 1 in info. I had initially passed a diagonal of all zeroes in, rather than making a row all zeroes.

swallan commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1672,23 +1675,39 @@ def test_gttrf_gttrs():
-        np.testing.assert_allclose(x, x_dgttrs, atol=atol)
-        np.testing.assert_allclose(b, b_cpy, atol=atol)
-        
+        np.testing.assert_array_equal(b, b_cpy)
```
I changed the wrapper on my machine so that b will be maintained, but I don't think it's implemented yet here.

Kai-Striega committed:
ENH: linalg: Allow ?gttrs to copy `b` matrix

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,69 @@ def test_getc2_gesc2():
+        b = A @ x
+
+        dgttrf = get_lapack_funcs('gttrf', dtype=dtype)
+        dgttrs = get_lapack_funcs('gttrs', dtype=dtype)
```
nitpick: Here the functions are named `dgttrf` and `dgttrs`, but they should be the more generic `gttrf` and `gttrs`. The reason for this is that LAPACK names are used to specify the function. The first letter represents the data type used. In this case **`d`**`gttrf` implies the function uses type `double` but this is not always correct as the test should change the type.

Also [`get_lapack_funcs`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.get_lapack_funcs.html) accepts a sequence of function names, so the above two lines could be squashed to:

```
gttrf, gttrs = get_lapack_funcs(('gttrf', 'gttrs'), dtype=dtype)
```

swallan replied:
Thanks for the feedback! I appreciate it. And, It's great to meet you as well.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,69 @@ def test_getc2_gesc2():
+
+        np.testing.assert_array_equal(dl, diag_cpy[0])
+        np.testing.assert_array_equal(d, diag_cpy[1])
+        np.testing.assert_array_equal(du, diag_cpy[2])
```
Here you don't have to write ``np.testing.assert_array_equal(...)`` in full. ``assert_array_equal`` is already imported at the start of the module in lines 12-14:

```python
from numpy.testing import (assert_equal, assert_array_almost_equal, assert_,
                           assert_allclose, assert_almost_equal,
                           assert_array_equal)
```

swallan replied:
Got it. Those were left over from running it in another file, I updated those and others to be the same style as the other test methods.

Kai-Striega reviewed (commented):
@swallan nice to finally meet you! 

Overall  your tests look good. I've left some review comments for you, these are mainly nitpicking/minor. I'll do a full review once @mdhaber and yourself are happy.

P.S. gttrs now lets you copy b.

mdhaber commented:
> I do think we should expose this as some kind of more specialised Exception Class

Yeah that sounds like a good idea!

swallan committed:
Merge branch 'lapack_support_gttrf_gttrs' of https://github.com/Kai-Striega/scipy into lapack_support_gttrf_gttrs

swallan committed:
TST: update tests for ?gttrf / ?gttrs

generalized gttrf/gttrs names, begin adding testing to regen A after using gttrf

swallan committed:
TST: ?gttrf / gttrs testing update

Singular matrix should be row of zeroes, not diagonal.

swallan committed:
TST: Add transposal test options for gttrs

this adds tests to ensure that trans=N, T, C work as expected within ?gttrs

swallan committed:
TST: add examples from nag manual as tests for ?gttrf/?gttrs

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
+        np.testing.assert_(_info != 0,
+                           "?gttrf: singular matrix ret info == 0")
+
+    du = np.array([2.1, -1.0, 1.9, 8.0])
```
Let's make this a separate test like `test_dttrf_NAG` so it only executes once (rather than once per data type). Also, please add a comment about the source of the test.

swallan replied:
Ok. My intent was to have it inside the test_gttrf_gttrs function, but outside the for loop. It might be indented one to many to be inside the for loop. Should I still make it a separate function?

mdhaber replied:
Oh, maybe your indentation is fine. I was thinking that the different datatypes were looped over by pytest, not an explicit loop in the function. 
Yes, I think it should be separated.
Thanks for dealing with the nitpicks. It's good to get this one "perfect" (according to me; then Kai can check) so we can just follow the example for the rest.

swallan replied:
Sounds good, I'm happy to deal with small things to make sure everything is exactly right. I separated it into a separate test function.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
+        with assert_raises(Exception):
+            gttrf(dl[0], d[:1], du[0])
+
+        transpose_options = ["N", "C", "T"]
```
We should probably just do "T" for real data types and "C" for complex data types.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
                                       x/scale, decimal=4)
+
+
+def test_gttrf_gttrs():
```
Please add comments explaining what each part tests.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
+
+def test_gttrf_gttrs():
+    np.random.seed(42)
+    for index, dtype in enumerate(DTYPES):
```
e.g. `# For each data type, use `?gttrf` and `?gttrs` to solve a random linear system`

swallan replied:
Will do!

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
+    np.random.seed(42)
+    for index, dtype in enumerate(DTYPES):
+        n = 10
+        if index % 2 == 0:
```
e.g. `#  set appropriate test tolerance for each data type`

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
+
+        _dl, _d, _du, du2, ipiv, info = gttrf(dl, d, du)
+
+        assert_array_equal(dl, diag_cpy[0])
```
e.g. `# confirm that copies are made / ?gttrf` does not modify inputs`

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
+            perm_fnl[i], perm_fnl[piv-1] = perm_fnl[piv-1], perm_fnl[i]
+
+        _A = np.diag(_dl, -1) + np.diag(_d, 0) + np.diag(_du, 1)
+        #assert_allclose(_A, A[perm_fnl], rtol=rtol)
```
e.g. `# check that the outputs of ?gttrf define an LU decomposition of A`

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,115 @@ def test_getc2_gesc2():
+        assert_array_equal(b, b_cpy)
+
+        assert_allclose(x, x_gttrs, rtol=rtol)
+
```
`# check that incompatible input sizes raise the appropriate exceptions`

swallan committed:
TST: Add #comments to explain tests, addition of proper LU composition

Kai-Striega commented:
My apologies for not being more active I fell sick and then focused on having some time off for Christmas. But I'll be more active again from now on.

Regarding the `overwrite_` arguments the documented behaviour is that:
> The original contents of the intent(in) argument may be altered by the Fortran/C function

so the argument **may**, not must, be overwritten by the function. I agree with that this doesn't need to be tested.

@mdhaber let me know when you would like me to do a more thorough review.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+            if index < 2:
+                Li = np.eye(n)
+            else:
+                Li = np.eye(n) + np.eye(n)*1j
```
I don't think there needs to be separate logic for complex case. The identity matrix is the identity matrix; there is not a distinct complex version.

mdhaber replied:
Ah, but it should be `Li = np.eye(n,dtype=dtype)`. Otherwise you lose the imaginary part below.
Actually, `dtype=dtype` should probably be used wherever you create an array.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+        temp = P[i,:].copy()
+        P[i,:] = P[piv, :]
+        P[piv,:] = temp
+
```
```suggestion
        L = L@P
```
I've actually never seen the last element of `ipiv` be something other than `n`, but I think in general the last permutation is needed.

mdhaber replied:
Actually, you can replace the creation of the last permutation matrix and `L=L@P` with:
```
i, piv = -1, ipiv[-1] - 1
L[:,[i,piv]] = L[:,[piv,i]]  # right multiply by final permutation matrix
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+        U = np.diag(_d, 0) + np.diag(_du, 1) + np.diag(du2, 2)
+        L = np.eye(n)
+
+        for i, m in enumerate(_dl):
```
Below, I wrote:
> It's horrendously inefficient multiplying all those zeros. As a matter of form, I'd suggest eliminating those explicit permutation matrix multiplications and replacing them with explicit column swaps.


What I meant is replace this loop with:

```
        for i, m in enumerate(_dl):
            # L is given in a factored form. See http://www.hpcavf.uclan.ac.uk/softwaredoc/sgi_scsl_html/sgi_html/ch03.html
            piv = ipiv[i] - 1
            L[:,[i,piv]] = L[:,[piv,i]]  # right multiply by permutation matrix
            L[:,i] += L[:,i+1]*m  # right multiply by factor Li, a rank-one modification of the identity
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+
+            L = L@P@Li
+
+        # one last permutation
```
The last permutation can be replaced with:
```
i, piv = -1, ipiv[-1] - 1
L[:,[i,piv]] = L[:,[piv,i]]  # right multiply by final permutation matrix
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+        else:
+            rtol = 10**-(np.finfo(dtype).precision-3)
+        # create the matrix in accordance with the data type
+        if index < 2:
```
let's replace the `if`-`else` logic by creating a function:
```
def my_rand(shape, dtype):
   x = np.random.rand(*shape) + np.random.rand(*shape)*1.0j
   return x.astype(dtype)
```
It will be useful here and in future PRs.
You'll need to filter the warning to avoid test failures.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+
+        # test that singular (row of all zeroes) matrix fails via info
+        dl[0] = 0
+        du[0] = 0
```
`dl[0]` is on the second row, so you don't need to set it to zero here.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+
+        # test that ?gttrf and ?gttrs work with transposal options
+        transpose_options = ["C", "T"]
+        for trans in transpose_options:
```
```suggestion

```
you don't need to do both transpose options for every type. `C` is for complex types, `T` is for real types. 
Also, let's move this up above the tests for inconsistent shapes.

swallan replied:
Got it.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+        # test that ?gttrf and ?gttrs work with transposal options
+        transpose_options = ["C", "T"]
+        for trans in transpose_options:
+            if trans == "T":
```
```suggestion
         if dtype in REAL_DTYPES:
            trans == "T"
            b_trans = A.T @ x
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+        # test that the result of ?gttrs matches the expected input
+        assert_allclose(x, x_gttrs, rtol=rtol)
+
+        # create matricies of inconsistent shapes
```
No need to create new arrays here. Just truncate the existing arrays.

swallan replied:
Ok!

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+
+        # test that ValueError is raised with incompatible matrix shapes
+        with assert_raises(ValueError):
+            gttrf(dl_misshaped, d, du)
```
For example,
```suggestion
            gttrf(dl[:-1], d, du)
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+    #%%
+    np.random.seed(42)
+    # use ?gttrf and ?gttrs to solve a random linear system for each dtype
+    for index, dtype in enumerate(DTYPES):
```
I know the examples all loop over the different dtypes, but I think we should [parameterize the test](https://docs.pytest.org/en/latest/parametrize.html).

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
                                       x/scale, decimal=4)
+
+
+def test_gttrf_gttrs():
```
```suggestion

@pytest.mark.parametrize("dtype", DTYPES)
def test_gttrf_gttrs(dtype):
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,158 @@ def test_getc2_gesc2():
+    for index, dtype in enumerate(DTYPES):
+        n = 10
+        # set appropriate test tolerance based on the data type
+        if index % 2 == 0:
```
Does this need to depend on the data type? I would prefer to avoid the `if` statement.

mdhaber replied:
Again, `tol=25*np.finfo(dtype).eps` appears elsewhere and it works fine here. And I think you can just use this number for both `rtol` and `atol`.

swallan replied:
I hadn't used `rtol=25*np.finfo(dtype).eps` here since it is a little too small to pass over the tests. For example, when `rtol` is `rtol=2.22045e-14`, an error is raised due to a max relative difference of `5.25248037e-14`. 

Right now I'll use `rtol=250*np.finfo(dtype).eps` so that `rtol` is` 5.5511...e-14`, unless you have another preference.

mdhaber replied:
I think they passed on my computer, so I thought it was fine. But an extra factor of 10 is fine, too.

mdhaber commented:
@swallan This is getting really close. I think this is the last round of major revisions. Afterward, I think the code will be much more concise and readable, and I think it will set us up well for unit tests of other functions. Thanks for your patience.

swallan committed:
TST: Add parametrization and make matrix permutations more efficient

removed if-else clauses, unneeded code, and corrected L U matrix permutation issues to work with all dtypes.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+    assert_array_equal(d, diag_cpy[1])
+    assert_array_equal(du, diag_cpy[2])
+
+    perm_fnl = np.arange(n)
```
`perm_fnl` is no longer used. Remove this and following lines.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+    d = my_rand(n, dtype=dtype)
+    dl = my_rand(n-1, dtype=dtype)
+
+    diag_cpy = np.array([dl.copy(), d.copy(), du.copy()])
```
This generates an object array - not simply a numerical array. Might as well just use a list.
```suggestion
    diag_cpy = [dl.copy(), d.copy(), du.copy()]
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+    for i, m in enumerate(_dl):
+        # L is given in a factored form. See http://www.hpcavf.uclan.ac.uk/softwaredoc/sgi_scsl_html/sgi_html/ch03.html
+        piv = ipiv[i] - 1
+        L[:,[i,piv]] = L[:,[piv,i]]  # right multiply by permutation matrix
```
Here and elsewhere we need space after commas (per PEP8).
```suggestion
        L[:, [i, piv]] = L[:, [piv, i]]  # right multiply by permutation matrix
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+    perm_fnl = np.arange(n)
+    for i, piv in enumerate(ipiv):
+        perm_fnl[i], perm_fnl[piv-1] = perm_fnl[piv-1], perm_fnl[i]
+
```
```suggestion
    # Generate L and U factors of A from return values of ?gttrf. 
    # L/U are initially lower/upper triangular by construction (and remain that way).
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+    L[:,[i,piv]] = L[:,[piv,i]]  # right multiply by final permutation matrix
+
+    _A = L @ U
+    # check that the outputs of ?gttrf define an LU decompisition of A
```
```suggestion
    # check that the outputs of ?gttrf define an LU decomposition of A
```

swallan replied:
👍

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+    gttrf, gttrs = get_lapack_funcs(('gttrf', 'gttrs'), dtype=dtype)
+
+    _dl, _d, _du, du2, ipiv, info = gttrf(dl, d, du)
+    # test to assure that the inputs of ?gttrf are maintained
```
```suggestion
    # test to assure that the inputs of ?gttrf are unmodified
```

swallan replied:
👍

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+                       "?gttrf: singular matrix ret info == 0")
+
+
+def my_rand(shape, dtype):
```
docstring
```suggestion
% generates a random matrix of desired data type
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+def test_gttrf_gttrs(dtype):
+
+    np.random.seed(42)
+    # use ?gttrf and ?gttrs to solve a random linear system for each dtype
```
I'd bring this comment to the top of the function and elaborate. For instance, you also check that return values of ?gttrf define LU matrices, etc... Summarize what this function tests.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+    # use ?gttrf and ?gttrs to solve a random linear system for each dtype
+    n = 4
+    # set appropriate test tolerance based on the data type
+    rtol = 250*np.finfo(dtype).eps
```
Let's put comment above inline
```suggestion
    rtol = 250*np.finfo(dtype).eps  # set test tolerance appropriate for data type
```

swallan replied:
👍

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,128 @@ def test_getc2_gesc2():
+
+    b_cpy = b.copy()
+    x_gttrs, info = gttrs(_dl, _d, _du, du2, ipiv, b)
+    # test that the inputs of ?gttrs are maintained
```
```suggestion
    # test that the inputs of ?gttrs are unmodified
```

mdhaber reviewed (changes requested):
Easy changes.

swallan committed:
TST: Clarify comments, remove unneeded lines, fix spelling typos.

swallan committed:
TST: Add URL explaining LU

mdhaber reviewed (approved):
@Kai-Striega I think this is ready for you.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+    d[0] = 0
+    __dl, __d, __du, _du2, _ipiv, _info = gttrf(dl, d, du)
+    np.testing.assert_(_info != 0,
+                       "?gttrf: singular matrix ret info == 0")
```
This test could be slightly more specific. The [?gttrf docs](http://www.netlib.org/lapack/explore-html/d1/d40/group__double_g_tcomputational_ga8ca64e542924cec56cbe9837b77d25b7.html) give the following description for ``info > 0``:

> if INFO = k, U(k,k) is exactly zero. The factorization has been completed, but the factor U is exactly             singular, and division by zero will occur if it is used to solve a system of equations.

Rather than the more general ``_info != 0`` we expect the first element (as you have set du[0] to exactly zero) to be zero, i.e. ``_info == 1``. The reason this is 1 is because we are switching between Fortran and C style indexing.

mdhaber replied:
OK, sounds like a good idea.

swallan replied:
So for this it looks like _info is not giving 1 for the `i`th element, but instead is `info = 10`, so not the first element of the matrix, but the very last?. (it gives 9 when n=9, 8 when n=8, etc).  Is this correct behavior?

Kai-Striega replied:
What is the value of `d[info - 1]` after the factorisation is complete? It could be that it has reordered the factors?

swallan replied:
I’ll have to come back to check on that tomorrow, but I committed the other changes for now!

swallan replied:
It looks like the last element of d from ?gttrf's output, `d[info-1],` is zero.

swallan replied:
So, does this mean that I should implement the test to check that the d[info-1]th value contains the illegal value, zero, that we put in d?

mdhaber replied:
Yes. After factorization, `__d[k]` is `U[k, k]`. Fortran (1-indexed) says `info=k=10`, so look in `d[info-1]` for the zero.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+def my_rand(shape, dtype):
+    # generates a random matrix of desired data type
+    x = np.random.rand(shape) + np.random.rand(shape)*1.0j
+    return x.astype(dtype)
```
On my machine, this raises a ``ComplexWarning`` which would fail any other tests that do not explicitly filter for these. If you want to avoid the warning, maybe try using the dtype as a conditional? e.g.:

```python
if dtype in COMPLEX_DTYPES:
    return (np.random.rand(shape) + np.random.rand(shape)*1.0j).astype(dtype)
return np.random.rand(shape).astype(dtype)
```
I haven't tested this, and it may need some tidying up but it should construct the matrix without raising a warning. Another way (which may be better) would be to ignore the warning *inside* the function. You can do this using [numpy.testing.suppress_warnings](https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.suppress_warnings.html)

mdhaber replied:
@swallan I think Kai's suggestion to do this with a conditional is good. Let's do that.

swallan replied:
👍

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+
+def test_gttrf_gttrs_NAG():
+    # test to assure that wrapper is consistent with NAG manual
+    # example problems
```
NAG gives two examples for ?gttrf; [f07cdf](https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07cdf.html) for real type and ~~f07cef~~ [f07crf](https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07crf.html) for complex types. However currently only f07cdf is tested.

The test also doesn't tell the user *which* examples are being tested. We know these, but should we expect a maintainer in  5 years time to know which examples we meant?

mdhaber replied:
Good call. We should add the complex test as a separate function, put the example number in the function name, and a link in the comment.

Kai-Striega replied:
I'm not sure if it's worth creating a whole extra function especially as the logic is identical. What I was doing with ?tbtrs is to parameterize the test with dtype. Then define the variables with an if/else. I know it's probably redundant to test each datatype, but the 4 tests require 7ms to run. 

Saying that, the test could also be parameterized with the arguments directly.

swallan replied:
I agree that links and example numbers in the name are a great idea, but I'm a bit confused about what you mean in that f07cdf is for real type and f07cef is for complex. The example from f07cdf is real and pertains to `dgttrf`,  f07cef is also real and pertains to `dgttrs`, and both of those are done in together in this test. 

If I'm missing something obvious on NAG examples about complex dtypes forgive me!

Kai-Striega replied:
My apologies, I copy and pasted the wrong identifier. The complex examples are [f07crf](https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07crf.html) and [f07csf](https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07csf.html). - I'll update my post with the correct links too.

swallan replied:
No problem–glad to see I wasn't missing something obvious.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+                       "?gttrf: singular matrix ret info == 0")
+
+
+def my_rand(shape, dtype):
```
The name ``my_rand`` doesn't tell me much about what the function actually does. From what I can see you're using it to generate a random matrix. I would suggest changing the name to be more descriptive of what the function is actually doing.

mdhaber replied:
Good call. Perhaps `generate_random_dtype_array`

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.filterwarnings("ignore:Casting complex")
+def test_gttrf_gttrs(dtype):
+    # The test uses ?gttrf and ?gttrs to solve a random system for each dtype,
+    # tests that the output of ?gttrf define LU matricies, that input
+    # parameters are unmodified, transposal options function correctly, that
+    # incompatible matrix shapes raise an error, and singular matrices return
+    # non zero info.
```
This looks good. The comments describe what is tested and the necessary information, but not more 👍

swallan replied:
👍

mdhaber commented:
Ok. Sam, please just parameterize this test rather than adding another function. It is up to you whether you'd like to add the inputs/outputs as parameters or get them inside the function based on dtype.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
+
+    _A = L @ U
+    # check that the outputs of ?gttrf define an LU decomposition of A
+    assert_allclose(A, _A, atol=rtol)
```
If we don't use `_A` anywhere else I think explicitly performing the LU (re?)-decomposition inside the assert will be just as clear eg:

```python
# check that the outputs of ?gttrf define an LU decomposition of A
assert_allclose(A,  L @ U, atol=rtol)
```

second the relative tolerance is used to set the absolute tolerance in ``assert_allclose(..., atol=rtol)`` is that meant to be rtol?

swallan replied:
Nice catch with the` atol=rtol` leftover. Due to some problems I discovered with using relative tolerance when comparing zeros, I think that atol did need to be set to something here.  I'll add a dtype dependent atol at the top for this.

swallan committed:
TST: parameterize NAG, clarify names and comments

This commit adds the complex NAG manual examples to the test along with the reals in a parameterized fashion. It also clarifies function names and adds comments to give sources and names of NAG examples. Lastly, fixes small typos.

swallan committed:
TST: ?gttrf refine _info check for illegal value

Kai-Striega commented:
@swallan I'm happy with the changes. I'll open this as a PR to the scipy branch.
